### ISSUE TITLE
What data should be collected from first pass (the "big compute")

### ISSUE THREAD

**Issue by rcedgar** [ababaian/serratus#42]
First pass through the data = map reads to Covid pan-genome. Which data should be collected and saved from this pass?

**Comment by rcedgar** [comment]
First pass through the data = map reads to Covid pan-genome. Which data should be collected and saved from this pass?

I think we should collect as much data as possible. It will be too hard and too slow to design the whole thing before starting. We can be looking at first pass results as they come in and use them to work on the second pass, as long as the first pass results have enough information.

The alignment score (AS: tag in the SAM file) is a much better confidence indicator than MAPQ. If we filter by a minimum AS threshold, then the poly-As and -Ts would be discarded without masking. This takes care of small fragments.

SAM (& BAM) is a very verbose format, we don't need most of the information. A more compact format such as PAF would be better. If <<1% of reads/study are hits, and the format is much more compact the SAM, can we store all the hits for every study? I would guess yes, in which case this is the best approach.

If we can't store all the hits, then we should store as much information as we can. In that case, the pipeline will look like this:

fastq-dump | bvfilter | bowtie2 | summarizer > summarydata

I can write the summarizer if we can agree what should go in there.

Small and simply summary: for each reference sequence R in the pan-genome, save the number of hits where R is the top hit for a few select ranges, e.g. AS is low, medium and high. Maximum size of this file is 3x the number of reference sequences, but would be much smaller for most studies because most Rs will have zero hits. If needed, we can use gzip or whatever. This is a simple solution which is robust against poly-X masking and discarding non-Cov sequences in the reference because we can do this in downstream analysis.

Even if we store all the hits, it would be nice to have summaries coming in so that we can get a quick overview of what the pipeline is finding. This can be done by having the summarizer write PAF records to stdout and a summary when it completes.

**Comment by ababaian** [comment]
Currently we are storing the following data from [this script](https://github.com/ababaian/serratus/blob/master/docker/serratus-merge/run_merge.sh).

- Aligned `bam` file. These are only the reads which are retained by `samtools view -b -G 12 -` for paired-end or `samtools view -b -F 4 - ` for unpaired. An important optimization here is to create a dummy bam-header that is smaller then the cov1r header (~500 kb per accession).
- The bam index `.bam.bai` file. This is not needed
- The `flagstat` file, output of  `samtools flagstat SRRXXXX.bam > SRRXXXX.flagstat`
- The `refCount` file which is a count of reads per sequence in `cov1r` generated by 
```
samtools view -h temporary.bam | \
  tee >(awk '/^[^@]/ {count[$3]++} END {for (id in count) {print id, count[id]}}' > refCount) | \
  samtools sort -@ $THREADS - | \
  samtools reheader 0.header.sam - >\
  merged_sorted.bam
```

**Comment by rcedgar** [comment]
paf.gz is ~3x smaller than bam. but if we can easily store bam, then sam/bam is more widely supported and allows conversion back to fastq which paf does not.

**Comment by ababaian** [comment]
Storage is not an issue in this case. We can eat the few TB of storage costs for a few months until we sort out the best archiving method. There's also the possibility of registering an AWS Open Data set which means we would get free archiving of the data.

The primary reason to retain these alignments is to convert back to fastq for first-pass assembly.

**Comment by rcedgar** [comment]
Sounds good,  will close this issue. Seems there will be "issues" as I ramp up the learning curve figuring out what you guys have already done, which is very impressive so far!
